File: solutions/haskell/raindrops/1/src/Raindrops.hs
module Raindrops (convert) where
import Control.Monad ((>>=))

convert :: Int -> String
convert n
 | res /= "" = res
 | otherwise = show n
   where
    res = [(3,"Pling"), (5, "Plang"), (7, "Plong")] >>= (\(m,s) -> if n `mod` m == 0 then s else "")

File: solutions/erlang/raindrops/1/src/raindrops.erl
-module(raindrops).

-export([convert/1, accPling/4, resOrInputStr/2]).

convert(N) ->
    Res = lists:foldl(
        fun({M,Str}, Acc) -> accPling(N,M,Acc,Str) end,
        "",
        [{3,"Pling"}, {5, "Plang"}, {7, "Plong"}]),
    resOrInputStr(Res,N).

accPling(N,M,Acc,Str) when N rem M =:= 0 -> Acc ++ Str;
accPling(_,_,Acc,_) -> Acc.

resOrInputStr(Res, _) when length(Res) > 0 -> Res;
resOrInputStr(_,   N) -> integer_to_list(N).
